lib/src/network/backends.rs

```diff
@@ -188,21 +201,142 @@ impl BackendList {
       })
   }
 
-  pub fn available_backends(&mut self) -> Vec<&mut Rc<RefCell<Backend>>> {
-    self.backends.iter_mut()
+  pub fn available_backends(&mut self) -> Vec<Rc<RefCell<Backend>>> {
+    self.backends.iter()
       .filter(|backend| (*backend.borrow()).can_open())
+      .map(|backend| (*backend).clone())
       .collect()
   }
 
-  pub fn next_available_backend(&mut self) -> Option<&mut Rc<RefCell<Backend>>> {
-    let mut backends:Vec<&mut Rc<RefCell<Backend>>> = self.available_backends();
-    if backends.is_empty() {
-      return None;
-    }
+  pub fn next_available_backend(&mut self) -> Option<Rc<RefCell<Backend>>> {
+    let backends = self.available_backends();
 
-    let rnd = random::<usize>();
-    let idx = rnd % backends.len();
+    if backends.len() == 0 {
+      None
+    } else {
+      self.load_balancing.next_available_backend(&backends)
+    }
+  }
 
-    Some(backends.remove(idx))
+  pub fn set_load_balancing_policy(&mut self, load_balancing_policy: LoadBalancingAlgorithms) {
+    match load_balancing_policy {
+      LoadBalancingAlgorithms::RoundRobin => self.load_balancing = Box::new(RoundRobinAlgorithm{ next_backend: 0 }),
+      LoadBalancingAlgorithms::Random => self.load_balancing = Box::new(RandomAlgorithm{}),
+    }
   }
 }
+
+#[cfg(test)]
+mod backends_test {
+
+  use super::*;
+  use std::{thread,sync::mpsc::*,net::TcpListener};
+
+
+  fn run_mock_tcp_server(addr: &str, stoper: Receiver<()>) {
+    let mut run = true;
+    let listener = TcpListener::bind(addr).unwrap();
+    listener.set_nonblocking(true).expect("Cannot set non-blocking");
+
+    thread::spawn(move || {
+      while run {
+        for stream in listener.incoming() {
+          // accept connections
+        }
+
+        if let Ok(()) = stoper.try_recv() {
+          run = false;
+        }
+      }
+    });
+  }
+
+  #[test]
+  fn it_should_retrieve_a_backend_from_app_id_when_backends_have_been_recorded() {
+    let mut backend_map = BackendMap::new();
+    let app_id = "myapp";
+
+    let backend_addr = "127.0.0.1:1236";
```

Comment:
please make the TCP ports coherent with the rest

Reply:
So it's 1235 the next value ?